.github/workflows/ci.yml

```diff
@@ -0,0 +1,220 @@
+name: GitHub CI
+
+on:
+  pull_request:
+  push:
+  schedule:
+    - cron: 0 10 * * Mon
+  workflow_dispatch:
+
+defaults:
+  run:
+    shell: 'bash -Eeuo pipefail -x {0}'
+
+jobs:
+  generate-matrix:
+    runs-on: ubuntu-22.04
+    outputs:
+      matrix-alpine: ${{ steps.set-matrix.outputs.matrix-alpine }}
+      matrix-ubuntu: ${{ steps.set-matrix.outputs.matrix-ubuntu }}
+      matrix-redhat: ${{ steps.set-matrix.outputs.matrix-redhat }}
+    steps:
+      - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+
+      - name: set-matrix
+        id: set-matrix
+        run: |
+          case "$GITHUB_EVENT_NAME" in
+            pull_request|push)
+              # use a small subset of modules to test against
+              COMMON_MODULES="base image-filter njs otel perl xslt"
+              ADD_MODULES="geoip"
+            ;;
+            schedule|workflow_dispatch)
+              # use all modules to test against
+              COMMON_MODULES="base auth-spnego brotli encrypted-session fips-check geoip2 headers-more image-filter lua ndk njs opentracing otel passenger perl rtmp set-misc subs-filter xslt"
+              ADD_MODULES="geoip"
+            ;;
+          esac
+
+          ALPINE="{\"include\":["
+          UBUNTU="{\"include\":["
+          REDHAT="{\"include\":["
+
+          for MODULE in $COMMON_MODULES; do
+              ALPINE="$ALPINE{\"module\":\"$MODULE\"},"
+              UBUNTU="$UBUNTU{\"module\":\"$MODULE\"},"
+              REDHAT="$REDHAT{\"module\":\"$MODULE\"},"
+          done
+
+          for MODULE in $ADD_MODULES; do
+              ALPINE="$ALPINE{\"module\":\"$MODULE\"},"
+              UBUNTU="$UBUNTU{\"module\":\"$MODULE\"},"
+          done
+
+          if [[ $ALPINE == *, ]]; then
+           ALPINE="${ALPINE%?}"
+          fi
+          ALPINE="$ALPINE]}"
+
+          if [[ $UBUNTU == *, ]]; then
+           UBUNTU="${UBUNTU%?}"
+          fi
+          UBUNTU="$UBUNTU]}"
+
+          if [[ $REDHAT == *, ]]; then
+           REDHAT="${REDHAT%?}"
+          fi
+          REDHAT="$REDHAT]}"
+
+          echo "matrix-alpine=$( echo "$ALPINE" )" >> $GITHUB_OUTPUT
+          echo "matrix-ubuntu=$( echo "$UBUNTU" )" >> $GITHUB_OUTPUT
+          echo "matrix-redhat=$( echo "$REDHAT" )" >> $GITHUB_OUTPUT
```

Comment:
I don't quite like building json using echo. It's a PITA to edit this code later. The following snippet does the same and either fail straight away or produce a valid json.
It's also shorter.
```
for os in alpine ubuntu redhat; do
    M=$COMMON_MODULES
    if [ "$os" != "redhat" ]; then
        M="$M $ADD_MODULES"
    fi
    jq -nRr '"matrix-" + $os +"=" + ( { include: [ $ARGS.positional[] | {module: . } ] } | tojson)' --arg os $os --args $M  >> $GITHUB_OUTPUT
done
```

Reply:
Implemented in the current patch version, thanks.